packages/nimble-components/src/patterns/error/specs/checkbox-radio-error.md

```diff
@@ -0,0 +1,103 @@
+
+![](spec-images/checkbox-no_error,no_wrap.PNG)
+
+![](spec-images/checkbox-no_error,wrap.PNG)
+
+![](spec-images/checkbox-error,no_wrap.PNG)
+
+![](spec-images/checkbox-error,wrap.PNG)
+
+![](spec-images/checkbox-long_error.PNG)
+
+![](spec-images/radio-horizontal,no_error.PNG)
+
+![](spec-images/radio-horizontal,error.PNG)
+
+![](spec-images/radio-vertical,error,no_wrap.PNG)
+
+![](spec-images/radio-vertical,error,wrap.PNG)
+
+![](spec-images/radio-vertical,long_error.PNG)
+
+### API
+
+#### Checkbox
+
+The `nimble-checkbox` will be updated to implement the `ErrorPattern` interface. This means it will have two new attributes:
+
+-   `error-text` - string
+-   `error-visible` - boolean
+
+The `error-visible` boolean will control whether or not a red exclamation icon will be visible next to the checkbox label and whether or not the `error-text` is displayed below the checkbox label.
+
+#### Radio Group
+
+The `nimble-radio-group` will be updated to implement the `ErrorPattern` interface. This means it will have two new attributes:
+
+-   `error-text` - string
+-   `error-visible` - boolean
+
+The `error-visible` boolean will control whether or not a red exclamation icon will be visible next to each radio label and whether or not the `error-text` is displayed below all the radio controls.
+
+While `error-visible` will be a new public attribute on the `nimble-radio-group`, it will also be added as an internal attribute on the `nimble-radio`, which will be responsible for displaying the error icon next to its label when that attribute is set. The `nimble-radio-group` will be responsible for keeping the `error-visible` state of each child `nimble-radio` instance synchronized with its own `error-visible` state. It will do this by updating `error-visible` on each `nimble-radio` any time its own `error-visible` state changes or any time its slotted radios changes.
+
+In order to accomplish these changes, the `nimble-radio-group` template and the `nimble-radio` template will both need to be forked from the FAST templates.
+
+#### Framework Support
+
+The Angular and Blazor wrappers for the `nimble-checkbox` and `nimble-radio-group` will be updated to include `error-text` and `error-visible`. The `nimble-radio` API will not be updated in Angular or Blazor to include `error-visible` since that is an internal attribute on the component.
+
+### Additional Work
+
+#### Checkbox and Radio styling updates
+
+Both the `nimble-checkbox` and `nimble-radio` need styling updates to align them with the current visual designs. These components have very similar styling needs, so there is likely opportunity to move much of the styles into a shared pattern. This will be considered during implementation.
+
+The styling changes include (but may not be limited to):
+
+-   Updating the control height to respect the `controlHeight` token rather than being sized to the text and/or checked/selected indicator
```

Comment:
When we leverage the token I want to encourage us to use it for default value configuration and not as part of calc expressions, etc. The intention is that normal css props like height are what we are configured off of and we aren't inventing a new api for sizing controls, we just have tokens for defaults. i.e. option 2 in this discussion https://github.com/ni/nimble/issues/610#issue-1277404188

Reply:
That's an interesting point. In my current prototype, I have one `calc` using the control height because the top/bottom padding of the control changes based on the control height. Therefore, that would need to change if we wanted to avoid any `calc` statements with the height, but the replacement css is alluding me at the moment.

Reply:
I vaguely remember similar concerns for the spinner and I think that is where this comes from:
https://github.com/ni/nimble/blob/b338c2500cfe0974feeb3b00eb3e5a92cffdf227/packages/nimble-components/src/spinner/styles.ts#L25

Don't think it blocks the HLD but fair warning (without really understanding the concerns) that we may end up deviating from pixel-perfect spacing to accommodate whole pixel alignment on grids to good simplifying approximations.